app/server/lib/OIDCConfig.ts

```diff
@@ -203,42 +247,105 @@ export class OIDCConfig {
+      response_types: ['code'],
+      ...extraMetadata,
+    });
+  }
+
+  private _forgeProtectionParamsForAuthUrl(protections: { codeVerifier?: string, state?: string, nonce?: string }) {
+    return _.omitBy({
+      state: protections.state,
+      nonce: protections.nonce,
+      code_challenge: protections.codeVerifier ?
+        generators.codeChallenge(protections.codeVerifier) :
+        undefined,
+      code_challenge_method: protections.codeVerifier ? 'S256' : undefined,
+    }, _.isUndefined);
+  }
+
   private async _generateAndStoreConnectionInfo(req: express.Request, targetUrl: string) {
     const mreq = req as RequestWithLogin;
     if (!mreq.session) { throw new Error('no session available'); }
-    const codeVerifier = generators.codeVerifier();
-    const state = generators.state();
-    mreq.session.oidc = {
-      codeVerifier,
-      state,
+    const oidcInfo: SessionObj['oidc'] = {
       targetUrl
     };
+    if (this.supportsProtection('PKCE')) {
+      oidcInfo.codeVerifier = generators.codeVerifier();
+    }
+    if (this.supportsProtection('STATE')) {
+      oidcInfo.state = generators.state();
+    }
+    if (this.supportsProtection('NONCE')) {
+      oidcInfo.nonce = generators.nonce();
+    }
 
-    return { codeVerifier, state };
+    mreq.session.oidc = oidcInfo;
+
+    return _.pick(oidcInfo, ['codeVerifier', 'state', 'nonce']);
   }
 
-  private async _retrieveCodeVerifierFromSession(req: express.Request) {
-    const mreq = req as RequestWithLogin;
+  private _buildEnabledProtections(section: AppSettings): EnabledProtectionsString[] {
+    const enabledProtections = section.flag('enabledProtections').readString({
+      envVar: 'GRIST_OIDC_IDP_ENABLED_PROTECTIONS',
+      defaultValue: 'PKCE,STATE',
+    })!.split(',');
+    if (enabledProtections.length === 1 && enabledProtections[0] === '') {
+      return [];
+    }
+    for (const protection of enabledProtections) {
+      if (!ENABLED_PROTECTIONS.hasOwnProperty(protection as EnabledProtectionsString)) {
+        throw new Error(`OIDC: Invalid protection in GRIST_OIDC_IDP_ENABLED_PROTECTIONS: ${protection}`);
+      }
+    }
+    return enabledProtections as EnabledProtectionsString[];
+  }
+
+  private async _retrieveChecksFromSession(mreq: RequestWithLogin):
+    Promise<{ code_verifier?: string, state?: string, nonce?: string }> {
     if (!mreq.session) { throw new Error('no session available'); }
+
+    const state = mreq.session.oidc?.state;
+    if (!state && this.supportsProtection('STATE')) {
+      throw new Error('Login or logout failed to complete');
+    }
+
     const codeVerifier = mreq.session.oidc?.codeVerifier;
-    if (!codeVerifier) { throw new Error('Login is stale'); }
-    return codeVerifier;
+    if (!codeVerifier && this.supportsProtection('PKCE')) { throw new Error('Login is stale'); }
+
+    const nonce = mreq.session.oidc?.nonce;
+    if (!nonce && this.supportsProtection('NONCE')) { throw new Error('Login is stale'); }
```

Comment:
In this case we want the user to login again ?

Reply:
Done in [eb8331b](https://github.com/gristlabs/grist-core/pull/883/commits/eb8331b647582f9335a0f921c997507bbdadf5f5)